cranelift-codegen/src/peephole.rs

```diff
@@ -0,0 +1,469 @@
+    fn move_reg(&mut self, src: RegUnit, dst: RegUnit) {
+        debug_assert!(self.map.len() == self.size);
+        self.map[dst as usize] = self.map[src as usize];
+        self.map[src as usize] = None;
+    }
+
+    // Does `env` have the exact binding characterised by `(ref, offs, size)` ?
+    fn has_exact_binding(&self, reg: RegUnit, offs: i32, size: u32) -> bool {
+        debug_assert!(self.map.len() == self.size);
+        if let Some(si) = &self.map[reg as usize] {
+            return si.offs == offs && si.size == size;
+        }
+        // No such binding.
+        false
+    }
+
+    // Does `env` have a binding characterised by `(offs, size)` but to a
+    // register, let's call it `reg2`, that isn't `reg`?  If so, return `reg2`
+    // and its associated SSA name.  Note that `reg2` will have the same bank
+    // as `reg`.  It is a checked error to call this function with a binding
+    // matching all three of `(reg, offs, size)`.
+    fn has_inexact_binding(
+        &self,
+        banks: &[RegBank],
+        reg: RegUnit,
+        offs: i32,
+        size: u32,
+    ) -> Option<(RegUnit, Value)> {
+        debug_assert!(self.map.len() == self.size);
+        // Find the range of RegUnit numbers for the bank that contains `ref`,
+        // and use that as our search space.  This is so as to guarantee that
+        // any match is restricted to the same bank as `reg`.
+        let (first_unit, n_units) = find_bank_limits(banks, reg);
+        for reg2 in first_unit..first_unit + n_units {
+            if let Some(si) = &self.map[reg2 as usize] {
+                if si.offs == offs && si.size == size {
+                    if reg2 == reg {
+                        panic!("has_inexact_binding: binding *is* exact!");
+                    }
+                    return Some((reg2, si.value));
+                }
+            }
+        }
+        // No such binding.
+        None
+    }
+
+    // Create the binding `(reg, value, offs, size)` in `env`, and throw away
+    // any previous binding associated with `reg`.
+    fn bind(&mut self, reg: RegUnit, value: Value, offs: i32, size: u32) {
+        debug_assert!(self.map.len() == self.size);
+        self.map[reg as usize] = Some(SlotInfo { value, offs, size });
+    }
+}
+
+impl<'a> Context<'a> {
+    /// Run the post-regalloc peephole optimiser.
+    fn cleanup(&mut self) {
+        let mut tot_regunits = 0;
+        for ru in self.reginfo.banks {
+            tot_regunits += ru.units;
+        }
+        // Perform redundant-fill removal.  We can visit the EBBs in any
+        // order, since there's no context carried across EBB boundaries at
+        // present.
+        for &ebb in self.domtree.cfg_postorder().iter().rev() {
+            self.cleanup_ebb(ebb, tot_regunits);
+        }
+    }
+
+    // Returns the regunit that `v` is allocated to.  Assumes that `v`
+    // actually resides in a register.
+    fn reg_of_value(&self, v: Value) -> RegUnit {
```

Comment:
In general, I think this is re-doing the work that's done by the RegDiversions data structure. In particular, this function and the function below are just the methods `reg` and `stack` from this data structure. Is there a way the RegDiversions data structure could be reused, to help with tracking?